src/lib/cpd/CpdEngine.ts

```diff
@@ -0,0 +1,215 @@
+import {xml2js} from 'xml-js';
+import checksum = require('checksum')
+import {Controller} from '../../Controller';
+import {Catalog, Rule, RuleGroup, RuleResult, RuleTarget, RuleViolation, TargetPattern} from '../../types';
+import {AbstractRuleEngine} from '../services/RuleEngine';
+import {Config} from '../util/Config';
+import {ENGINE, Severity} from '../../Constants';
+import {PmdCatalogWrapper} from '../pmd/PmdCatalogWrapper';
+import * as engineUtils from '../util/CommonEngineUtils';
+import CpdWrapper from './CpdWrapper';
+
+
+export class CpdEngine extends AbstractRuleEngine {
+
+	public static ENGINE_ENUM: ENGINE = ENGINE.CPD;
+	public static ENGINE_NAME: string = ENGINE.CPD.valueOf();
+
+	private logger: Logger;
+	private config: Config;
+	private initialized: boolean;
+	private cpdCatalog: Catalog;
+	private languageToPathMap;
+	protected pmdCatalogWrapper: PmdCatalogWrapper;
+
+	public getName(): string {
+		return CpdEngine.ENGINE_NAME;
+	}
+
+	public getTargetPatterns(): Promise<TargetPattern[]> {
+		return this.config.getTargetPatterns(ENGINE.CPD);
+	}
+
+	public getCatalog(): Promise<Catalog>{
+		return Promise.resolve(this.cpdCatalog);
+	}
+
+	/* eslint-disable-next-line no-unused-vars, @typescript-eslint/no-unused-vars */
+	public shouldEngineRun(ruleGroups: RuleGroup[], rules: Rule[], target: RuleTarget[], engineOptions: Map<string, string>): boolean {
+		return true;
+	}
+	
+	// eslint-disable-next-line @typescript-eslint/no-unused-vars
+	public async run(ruleGroups: RuleGroup[], rules: Rule[], targets: RuleTarget[], engineOptions: Map<string, string>): Promise<RuleResult[]> {
+		try {
+			const results = []
+			// cpd only allows one language to be speificed at a time, so we must run it on each (apex and vf)
+			for (const language of this.cpdCatalog.rules[0].languages) {
+				const targetPaths: string[] = [];
+				for (const target of targets) {
+					for (const targetPath of target.paths) {
+						// when using --filelist, CPD will try to parse files not associated with language
+						// for apex, we only want to add .cls and .trigger files and for vf, we only want to look at .page and .component files
+						if (this.validateFileType(language, targetPath)) {
+							targetPaths.push(targetPath);
+						}
+					}
+				}
+
+				if (targetPaths.length === 0) {
+					this.logger.trace(`No matching cpd ${language} target files found. Nothing to execute.`);
+					return [];
+				}
+				this.logger.trace(`About to run CPD (${language}). Targets: ${targetPaths.length}`);
+
+				const selectedTargets = targetPaths.join(',');
+				const stdout = await CpdWrapper.execute(selectedTargets, language, await this.config.getMinimumTokens(ENGINE.CPD));
+				results.push(...this.processStdOut(stdout));
+				
+				this.logger.trace(`Found ${results.length} for CPD`);
+			}   
+			return results;
+		} catch (e) {
+			this.logger.trace('Cpd evaluation failed: ' + (e.message || e));
+			throw new SfdxError(e.message || e);
```

Comment:
If the first language fails, the second language wouldn't have a chance to run. What do you think about handling this exception per language iteration?